crates/accelerate/src/sabre/route.rs

```diff
@@ -0,0 +1,584 @@
+                Some(NodeIndex::new(qubits[1].index())),
+                |_| Ok(1.),
+                Some(&mut shortest_paths),
+            ) as PyResult<Vec<Option<f64>>>)
+                .unwrap();
+            shortest_paths
+                .get(&NodeIndex::new(qubits[1].index()))
+                .unwrap()
+                .iter()
+                .map(|n| PhysicalQubit::new(n.index() as u32))
+                .collect::<Vec<_>>()
+        };
+        // Insert greedy swaps along that shortest path, splitting them between moving the left side
+        // and moving the right side to minimise the depth.  One side needs to move up to the split
+        // point and the other can stop one short because the gate will be routable then.
+        let split: usize = shortest_path.len() / 2;
+        current_swaps.reserve(shortest_path.len() - 2);
+        for i in 0..split {
+            current_swaps.push([shortest_path[i], shortest_path[i + 1]]);
+        }
+        for i in 0..split - 1 {
+            let end = shortest_path.len() - 1 - i;
+            current_swaps.push([shortest_path[end], shortest_path[end - 1]]);
+        }
+        current_swaps.iter().for_each(|&swap| self.apply_swap(swap));
+        closest_node
+    }
+
+    /// Return the swap of two virtual qubits that produces the best score of all possible swaps.
+    fn choose_best_swap(&mut self) -> [PhysicalQubit; 2] {
+        self.swap_scratch.clear();
+        let mut min_score = f64::MAX;
+        // The decay heuristic is the only one that actually needs the absolute score.
+        let dist = &self.target.distance;
+        let absolute_score = match self.heuristic {
+            Heuristic::Decay => {
+                self.front_layer.total_score(dist)
+                    + EXTENDED_SET_WEIGHT * self.extended_set.total_score(dist)
+            }
+            _ => 0.0,
+        };
+        for swap in obtain_swaps(&self.front_layer, self.target.neighbors) {
+            let score = match self.heuristic {
+                Heuristic::Basic => self.front_layer.score(swap, dist),
+                Heuristic::Lookahead => {
+                    self.front_layer.score(swap, dist)
+                        + EXTENDED_SET_WEIGHT * self.extended_set.score(swap, dist)
+                }
+                Heuristic::Decay => {
+                    self.qubits_decay[swap[0].index()].max(self.qubits_decay[swap[1].index()])
+                        * (absolute_score
+                            + self.front_layer.score(swap, dist)
+                            + EXTENDED_SET_WEIGHT * self.extended_set.score(swap, dist))
+                }
+            };
+            if score < min_score - BEST_EPSILON {
+                min_score = score;
+                self.swap_scratch.clear();
+                self.swap_scratch.push(swap);
+            } else if (score - min_score).abs() < BEST_EPSILON {
+                self.swap_scratch.push(swap);
+            }
```

Comment:
I've been diving into the swap selection logic in the routing algorithm and something caught my attention regarding how we handle scoring ties. From my experience testing Sabre on various circuits, these ties seem to be a common occurrence. It seems to me that the order of iteration of `obtain_swaps` is deterministic, so we're consistently picking the first optimal swap we come across. If this is true, then we might be inadvertently biasing our selection towards certain swaps, potentially missing out on exploring other equally good options that could prove beneficial, particularly in more complex or larger circuits. Currently, it does not seem like we have a tie-resolution mechanism. We could consider some randomness in dealing with ties. This could get a broader view of the optimal paths across different runs.

Reply:
- "ties seem to be a common occurrence": yep, they are.
- "the order of iteration of `obtain_swaps` is deterministic": yes, 100% - we need this for determinism of the algorithm for a fixed seed.
- "we're consistently picking the first optimal swap we come across": this is where you're going a bit wrong, I think.  We don't pick the first optimal one; we track our current "optimal score" (up to some slightly fuzzy epsilon), and whenever we see a swap that's as good as the best, we push it into a `Vec` (same as a Python `list`, for all intents and purposes) (line 368).  If we find a new best, we reset that tracking (lines 365 and 366). Once we've examined all swaps, we use the pRNG to randomly choose between all swaps that were the best (line 371, just below this diff) - that's where the random seed comes in, and in almost all circuits, that's the only use of randomness.

So our "tie-resolution mechanism" is just "pick randomly from all swaps that were within epsilon of the best seen", and in the preset pass managers, we run Sabre in parallel with quite a few different random seeds to try and get a view of the optimal paths.  Was there some other part you were thinking of?

Fwiw, I'm also interested in trialling a "best swap" chooser where we assign _all_ swaps a probability of being chosen (for some loss function that I've not had any particularly bright ideas about) and then choose from that, rather than only looking at the best swaps.

Reply:
Oh yes that is correct and thank you for the clarification. And no, I wasn’t thinking of another specific part of the algorithm for tie-resolution. Your explanation clarified my misunderstanding about how ties are handled within the current mechanism. Also I found the concept of experimenting with a probabilistic "best swap" chooser quite promising. I had dabbled with ideas for a loss function in this context before, but my focus shifted elsewhere. It seems like a worthwhile direction for further exploration, and I will try to revisit those preliminary thoughts.